polar-core/src/vm.rs

```diff
@@ -2137,11 +2113,15 @@ impl PolarVirtualMachine {
                 return self.push_goal(applicable);
             }
 
-            // Try to unify the arguments with renamed parameters.
-            // TODO: Think about using backtrack so that we don't
-            // leave temporary bindings around.
-            let args = self.rename_vars(args.clone());
+            // We need to check each argument for applicability,
+            // but unwind any bindings made during the check. So
+            // we will backtrack to this choice point on success.
+            self.push_choice(vec![vec![applicable]]);
```

Comment:
I'm struggling to get enough context to reason about this without reading through the full method again. What does the choice stack look like at the end of this method?
Something like:

[
- check_applicable goals (unify goals followed by backtrack, so backtracks *if successful*)
- inapplicable rules - has a cut so applicable rules doesn't get executed afterwards?

], 
[
 - applicable rules <- this is pushed onto the "current" choice stack so is an alternative after the head (here represented above) is completed

]

Reply:
Feels like a double negative somehow? Why isn't the choice point: [check_applicable + backtrack + cut + applicable, inapplicable] for example?

Reply:
Wait, backtrack would clear the other goals wouldn't it. Hmm, I guess that makes sense. The use of `push_choice` feels a little too like manual stack manipulation to me.

Reply:
Also, if the applicable check succeeds, doesn't this mean we still would try the inapplicable branch too?

Reply:
No, it's cut. But I will work on a better comment or diagram to explain the admittedly confusing flow control.

Reply:
Oh I see, I missed https://github.com/osohq/oso/pull/420/files/433da70ca8ee5acac745276a73dbccfce5e72995#diff-99e04680f3976c570014561774238bdeR2138